### PR TITLE
[F3D] Fog Export Fix

### PR DESCRIPTION
**Pull request by jesusyoshi54**
Fog now exports only when the `set_fog` boolean is checked and fog is used in the material, which is determined if `g_fog` is enabled, or if fog is an input to any of the blend variables. Previous behavior only checked if `set_fog` was on, which would mistakenly allow fog to be exported when the `set_fog` checkbox was invisible (and left checked from prev config) when material was changed to no longer have any fog inputs.

### CODE REVIEW

**Comment by Lilaa3** [comment]
I don´t know if I like this, one could use the alpha in the combiner with a texedge render mode for example, and this would break that ability. Also I don't see the behavior you are describing of fog being hidden in the UI?

**Comment by jesusyoshi54** [comment]
> I don´t know if I like this, one could use the alpha in the combiner with a texedge render mode for example, and this would break that ability.

I'm not sure what will be broken here.

> Also I don't see the behavior you are describing of fog being hidden in the UI?
Then I must be dumb and mistaken. I could have sworn this was an issue, but it might have been fixed sometimes with the render mode changes sauraen made.

**Comment by Lilaa3** [comment]
What will be broken is that, since the material is technically not using fog in render mode, fast64 will just not export the actual fog

**Comment by jesusyoshi54** [comment]
If fog isn't used in the render mode then why would it matter if it is unset.
If you're using shade alpha in the combiner you need g_fog enabled, in which case it will export.

**Comment by Dragorn421** [comment]
Hm I'd like to point out the combiner may use shade alpha without intending for fog, with said alpha being provided by the user in the Vtx data (alongside either verter colors or normals)

**Comment by Lilaa3** [comment]
Ah sorry I misread this, I thought you were checking for both g_fog and a fog render mode, my bad, and also set fog does get hidden still, sorry for so much confusion on my part

File: fast64_internal/f3d/f3d_material.py [hunk 1]
```diff
@@ -3231,6 +3209,29 @@ class RDPSettings(PropertyGroup):
...
+        return (getattr(self, f"blend_{val}") for val in ("p1", "p2", "m1", "m2"))
+
+    @property
+    def blend_alpha_inputs(self):
+        return (getattr(self, f"blend_{val}") for val in ("a1", "a2", "b1", "b2"))
+
+    @property
+    def blend_inputs(self):
+        return self.blend_color_inputs
+        return self.blend_alpha_inputs
```
Comment (Lilaa3) [inline fast64_internal/f3d/f3d_material.py:3227]: Did you mean to do something like
```suggestion
        return self.blend_color_inputs + self.blend_alpha_inputs
```

File: fast64_internal/f3d/f3d_material.py [hunk 2]
```diff
@@ -3231,6 +3209,29 @@ class RDPSettings(PropertyGroup):
         update=update_node_values_with_preset,
     )
 
+    @property
+    def using_fog(self):
+        return self.g_fog or any(["FOG" in input for input in self.blend_inputs])
```
Comment (Lilaa3) [inline fast64_internal/f3d/f3d_material.py:3214]: I personally would just check for the two blend modes directly for the sake of clarity, color inputs will never get new inputs added there is just those two for fog so its good to make that explicit I think
```suggestion
        return self.g_fog or self.does_blender_use_input("G_BL_CLR_FOG") or self.does_blender_use_input("G_BL_A_FOG")
```

File: fast64_internal/f3d/f3d_material.py [hunk 3]
```diff
@@ -3231,6 +3209,29 @@ class RDPSettings(PropertyGroup):
...
+    @property
+    def blend_inputs(self):
+        return self.blend_color_inputs
+        return self.blend_alpha_inputs
+
+    def does_blender_use_input(self, setting: str) -> bool:
+        for input in self.blend_inputs:
+            if setting == input:
+                return True
+        return False
```
Comment (Lilaa3) [inline fast64_internal/f3d/f3d_material.py:3233]: this logic could be one line
```suggestion
        return any(input == setting for input in self.blend_inputs)
```

**Review by Lilaa3 — CHANGES_REQUESTED** [review]
I really like the way you implemented these changes, Just a few small details

File: fast64_internal/f3d/f3d_material.py (hunk 1 above)
Comment (jesusyoshi54) [reply fast64_internal/f3d/f3d_material.py:3227]: These are generators so I did mean that

**Review by Lilaa3 — APPROVED** [review]
The code works perfectly, I like how it is written, merge soon.

**Pull request merged**